docs: section about custom source and sink types (#202)
eclipse-edc/docs: developer/wip/for-contributors/data-plane/data-plane-extensions.md

# Data Plane extensions

<!-- TOC -->
* [Data Plane extensions](#data-plane-extensions)
  * [1. The `DataPlaneManager`](#the-dataplanemanager`)
    * [1.1 Consumer PULL](#consumer-pull-flow)
    * [1.2 Provider PUSH](#provider-push-flow)
  * [2. The Data Plane Framework ](#the-data-plane-framework)
    * [2.1 Transfer Service](#transferservice)
    * [2.2 Pipeline Service](#pipelineservice)
  * [3. Writing custom Source/Sink ](#writing-custom-sourcesink)
    * [3.1 Custom DataSource ](#custom-source)
    * [3.2 Custom DataSink ](#custom-datasink)
    * [3.3 Executing the transfer ](#executing-the-transfer)

<!-- TOC -->

> The EDC Data Plane is a component responsible for transmitting data using a wire protocol and can be easily extended using the [DFF](#the-data-plane-framework) for supporting different protocols and transfer types. 

Reviewer: ```suggestion
The EDC Data Plane is a component responsible for transmitting data using a wire protocol and can be easily extended using the [Data Plane Framework (DPF)](#the-data-plane-framework) for supporting different protocols and transfer types. 
```

The main component of an EDC data plane is the [DataPlaneManager](#the-dataplanemanager).

> ## The DataPlaneManager

Reviewer: the ToC already has the proper numbering, but the headings don't

The `DataPlaneManager` manages execution of data plane requests, using the EDC [State Machine](../control-plane/programming-primitives.md#1-state-machines) pattern for tracking the state of data transmissions. 

> It receives `DataFlowStartMessage` from the [Control Plane](../contributor-handbook.md#2-the-control-plane) through [data plane signaling](./data-plane-signaling/data-plane-signaling.md) protocol if it's deployed as standalone process, or directly with method call when it's embedded in the same process.

Reviewer: ```suggestion
It receives `DataFlowStartMessage` from the [Control Plane](../contributor-handbook.md#2-the-control-plane) through the [data plane signaling](./data-plane-signaling/data-plane-signaling.md) protocol if it's deployed as standalone process, or directly via method call when it's embedded in the same process.
```

### TransferService

Given a `DataFlowStartMessage`, an implementation of a `TransferService` can transfer data from a `sourceDataAddress` to a `destinationDataAddress`. 

The `TransferService` does not specify how the transfer should happen. It can be internally in the data plane or it could leverage on external and more performant system.

Relevant methods of the `TransferService` are:

```java
public interface TransferService {

    boolean canHandle(DataFlowStartMessage request);

    Result<Boolean> validate(DataFlowStartMessage request);

    CompletableFuture<StreamResult<Object>> transfer(DataFlowStartMessage request);
}
```

- The `canHandle` expresses if the `TransferService` implementation is able to fulfill the transfer request expressed in the `DataFlowStartMessage`.

- The `validate` performs a validation on the content of a `DataFlowStartMessage`. 

- The `transfer` triggers a data transfer from a `sourceDataAddress` to a `destinationDataAddress`.

An implementation of a `TransferService` bundled with the [DPF](#the-data-plane-framework) is the [PipelineService](#pipelineservice).

### PipelineService

> The `PipelineService` is an extension of [TransferService](#transferservice) that leverages on internal Data-Plane transfer mechanism.

Reviewer: ```suggestion
The `PipelineService` is an extension of [TransferService](#transferservice) that leverages onaninternal Data-Plane transfer mechanism.
```

```java
public class FileDataSource implements DataSource {

    private final File sourceFile;

    public FileDataSource(File sourceFile) {
        this.sourceFile = sourceFile;
    }

    @Override
    public StreamResult<Stream<Part>> openPartStream() {
        return StreamResult.success(Stream.of(new FileStreamPart(sourceFile)));
    }

    @Override
    public void close() {
    }

    private record FileStreamPart(File file) implements Part {

        @Override
        public String name() {
            return file.getName();
        }

        @Override
        public InputStream openStream() {
            try {
                return new FileInputStream(file);
            } catch (FileNotFoundException e) {
                throw new RuntimeException(e);
            }
        }
    }
}
```

The relevant method is the `openPartStream`, which will be called for plumbing the source and sink. The `openPartStream` returns a `Stream` of `Part`s, as the `DataSource` can be composed by more that one part (e.g. folders). The `openPartStream` does not actually opens a Java `InputStream`, but rather returns a just stream of `Part`s.

> How to actually transform a `Part` into an `InputStream` must be implemented. In our case the `FileStreamPart#openStream` just returns a `FileInputStream` from the input `File`.

Reviewer: ```suggestion
Transforming a `Part` into an `InputStream` is the main task of the `DataSource` implementation. In our case the `FileStreamPart#openStream` just returns a `FileInputStream` from the input `File`.
```
